src/vcpkg/download.cpp

```diff
@@ -0,0 +1,77 @@
+#include <vcpkg/base/fmt.h>
+#include <vcpkg/base/jsonreader.h>
+#include <vcpkg/base/strings.h>
+
+#include <vcpkg/download.h>
+
+namespace vcpkg
+{
+    struct GitHubDeserializer final : Json::IDeserializer<DownloadedFile>
+    {
+        LocalizedString type_name() const override { return LocalizedString::from_raw("GitHub"); }
+
+        Optional<type> visit_object(Json::Reader& r, const Json::Object& obj) const override
+        {
+            DownloadedFile result;
+
+            result.kind = DownloadType::GitHub;
+            result.headers.emplace_back("Accept: application/vnd.github+json");
+            result.headers.emplace_back("X-GitHub-Api-Version: 2022-11-28");
+
+            std::string repo;
+            r.required_object_field(type_name(), obj, "repo", repo, Json::UntypedStringDeserializer::instance);
+            std::string ref;
+            r.required_object_field(type_name(), obj, "ref", ref, Json::UntypedStringDeserializer::instance);
+            r.required_object_field(type_name(), obj, "sha512", result.sha_512, Json::UntypedStringDeserializer::instance);
+            r.required_object_field(type_name(), obj, "out-var", result.out_var, Json::UntypedStringDeserializer::instance);
+            std::string gh_gost = "github.com";
+            r.optional_object_field(obj, "host", gh_gost, Json::UntypedStringDeserializer::instance);
+            std::string auth_token;
+            if (r.optional_object_field(obj, "authorization-token", auth_token, Json::UntypedStringDeserializer::instance))
+            {
+                result.headers.push_back("Authorization: Bearer " + auth_token);
+            }
+
+            r.optional_object_field(obj, "patches", result.patches, Json::IdentifierArrayDeserializer::instance);
+
+            result.urls.emplace_back(fmt::format("https://api.{}/repos/{}/tarball/{}", gh_gost, repo, ref));
```

Comment:
I decided to build the URL here in order to unify and simplify `DownloadedFile` for all downloading methods.